serverutils: provide SQLConn/SQLConnE in ApplicationLayerInterface

knz opened the pull request:
Fixes  #107672.
Part of solving #107058.
Informs #106772.

Epic: CRDB-18499

knz committed:
serverutils: provide SQLConn/SQLConnE in ApplicationLayerInterface

This simplifies test code and prevents common mistakes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107659)
<!-- Reviewable:end -->

knz commented:
TODO: remove StorageClusterConn

knz committed:
testserver: new methods `SQLConnForUser` and `SQLConnForUserE`

This deprecates the following pattern:
```go
testUserURL, cleanupFn := sqlutils.PGUrl(t,
	s.AdvSQLAddr(), "dbname", "username")
defer cleanupFn()
testUserDb, err := gosql.Open("postgres", testUserURL.String())
require.NoError(t, err)
defer testUserDb.Close()
```

Which can now be replaced by:
```
testUserDb := s.ApplicationLayer().SQLConnForUser(t, "username", "dbname")
```

Release note: None

knz committed:
testcluster: remove `StorageClusterConn`

This is replaced by `.SystemLayer(0).SQLConn()`.

Release note: None

stevendanna commented on pkg/server/testserver_sqlconn.go:
```diff
@@ -0,0 +1,115 @@
+	stopper.AddCloser(stop.CloserFn(func() {
+		err := goDB.Close()
+		cleanupFn()
+		log.Infof(context.Background(), "closed test SQL connection: %v", err)
```
Perhaps only log this on error. Up to you.

knz replied:
Done.

stevendanna reviewed (approved):
Nice. 

I reviewed the testserver/testcluster changes, the backupccl changes, and then skimmed the other packages.

knz commented:
bors r=stevendanna

stevendanna commented:
We should consider backporting this or it will be hard to backport other test-related fixes.

yuzefovich reviewed (commented):
Nice cleanup!

> Reviewed 4 of 30 files at r1, 2 of 50 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107659#-:-NaT6G9kAaDPnveqZHzI:b7oy6or)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
